### PR TITLE
Add custom logic for X509 authentication using SAN

### PR DESCRIPTION
**Pull request by narendly**
This commit adds an optional logic path that allows the ZK server to populate ServerCnxn.authInfo using a field from SAN instead of using the vanilla Subject DN of the client x509 certificate. 4 JVM property fields have been added and can be used to enable SAN-based authentication using regex for matching and extracting from the matched SAN field.

A test case has been added to X509AuthTest that specifically tests the SAN-based client ID extraction.

### CODE REVIEW

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java [hunk 1]
```diff
@@ -51,7 +57,18 @@
 public class X509AuthenticationProvider implements AuthenticationProvider {
```
Comment (abhilash1in) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java:57]: We're doing quite a few customizations to `X509AuthenticationProvider` and can lead to merge conflicts in the future. Would it make more sense to create a new AuthenticationProvider (or maybe inherit from `X509AuthenticationProvider`) with all these customizations?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java [hunk 2]
```diff
@@ -51,7 +57,18 @@
...
+    /**
+     * The following System Property keys are used to extract clientId from the client cert.
+     * @see #getClientId(X509Certificate)
+     */
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDTYPE = "zookeeper.X509AuthenticationProvider.clientCertIdType";
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANMATCHTYPE = "zookeeper.X509AuthenticationProvider.clientCertIdSanMatchType";
+    // Match Regex is used to choose which entry to use
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANMATCHREGEX = "zookeeper.X509AuthenticationProvider.clientCertIdSanMatchRegex";
+    // Extract Regex is used to construct a client ID (acl entity) to return
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANEXTRACTREGEX = "zookeeper.X509AuthenticationProvider.clientCertIdSanExtractRegex";
```
Comment (abhilash1in) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java:69]: Just curious, can these be set via zoo.cfg or are they set only via system properties? Setting these via zoo.cfg is preferable.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java [hunk 3]
```diff
@@ -178,9 +195,82 @@ public String getScheme() {
...
+
+        // Extract a substring from the found match using extractRegex
+        Pattern extractPattern = Pattern.compile(extractRegex);
+        Matcher matcher = extractPattern.matcher(matched.iterator().next().get(1).toString());
+        if (matcher.find()) {
+            String result = matcher.group();
+            LOG.info("X509AuthenticationProvider::matchAndExtractSAN(): returning extracted "
+                + "client ID: {}", result);
+            return result;
+        } else {
```
Comment (junkaixue) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java]: NIT: no need for else clause when if has return value to terminate function.

**Review by junkaixue — APPROVED** [review]
overall, lgtm

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java (hunk 2 above)
Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java:69]: Both work. For zoo.cfg, please remove the `zookeeper.` prefix.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java (hunk 1 above)
Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java:57]: This isn't that much customization considering both logic paths are based on X509 Certificates. This way, we avoid boilerplate. We will consider a different implementation if we consider it necessary (for example, another auth mechanism that integrates with a 3rd party library/service or something that uses more than just X509 certs).

**Review by abhilash1in — APPROVED** [review]
LGTM

**Pull request merged**